OpenTripPlanner technote (#617)
ropensci/roweb2: content/technotes/2020-02-14-opentripplanner.Rmd

---
slug: "opentripplanner"
title: "opentripplanner: Fast and easy multimodal trip planning in R with OpenTripPlanner"
package_version: 0.2.0
date: 2020-02-22
authors:
  - Malcolm Morgan
categories:
  - technotes
topicid: 
tags:
- R
- transport
- spatial
- geospatial
- routing
output: 
  html_document:
    keep_md: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.path = "", comment = "")

knitr::knit_hooks$set(
  plot = function(x, options) {
    hugoopts <- options$hugoopts
    paste0(
      "{{<figure src=",
      '"', x, '" ',
      if (!is.null(hugoopts)) {
        glue::glue_collapse(
          glue::glue('{names(hugoopts)}="{hugoopts}"'),
          sep = " "
        )
      },
      ">}}\n"
    )
  }
)
```

> With services like Google Maps, finding the fastest route from A to B has become quick, cheap, and easy. Not just for driving but walking, cycling and public transport too. But in the field of transport studies, we often want not only a single route, but thousands or millions of routes. This is where we hit a problem for services such as Google or the [Open Route Service](https://openrouteservice.org/), usually only allow a limited number of free routes per day (typically around 1000). So routing is either very time consuming or expensive. Another problem is that we may not be interested in the current travel options, but how those options may change in the future. Such as, after a new bridge has been built or a new bus timetable has been introduced. Therefore, researchers can find it useful to run their own routing services where they have more control and can produce as many routes as necessary. 

Reviewer: ```suggestion
With services like Google Maps, finding the fastest route from A to B has become quick, cheap, and easy. Not just for driving but walking, cycling and public transport too. But in the field of transport studies, we often want not only a single route, but thousands or millions of routes. This is where we hit a problem for services such as Google or the [Open Route Service](https://openrouteservice.org/), that usually only allow a limited number of free routes per day (typically around 1000). So routing is either very time consuming or expensive. Another problem is that we may not be interested in the current travel options, but how those options may change in the future. Such as, after a new bridge has been built or a new bus timetable has been introduced. Therefore, researchers can find it useful to run their own routing services where they have more control and can produce as many routes as necessary. 
```

added "that", before "usually only allow"

> The aim of the OpenTripPlanner for R package[^1] to make it easy to set up and use a multimodal trip planner in R. [OpenTripPlanner](https://www.opentripplanner.org/) (OTP) is an open-source multimodal trip planner written in Java. It uses OpenStreetMap for walking, cycling, and driving directions and uses [GTFS](https://developers.google.com/transit/gtfs) files for public timetables. The R package makes it easy to set up and use OpenTripPlanner on your local computer or connecter to a sever running OpenTripPlanner.

Reviewer: ```suggestion
The aim of the OpenTripPlanner for R package[^1] to make it easy to set up and use a multimodal trip planner in R. [OpenTripPlanner](https://www.opentripplanner.org/) (OTP) is an open-source multimodal trip planner written in Java. It uses OpenStreetMap for walking, cycling, and driving directions and uses [GTFS](https://developers.google.com/transit/gtfs) files for public timetables. The R package makes it easy to set up and use OpenTripPlanner on your local computer or connect to a server running OpenTripPlanner.
```

"connecter to a sever" -> "connect to a server"

```{r, eval=FALSE} 
path_data <- file.path(tempdir(), "OTP") # Make a folder to store the data 
dir.create(path_data)  
path_otp <- otp_dl_jar(path_data)        # Download the OTP 
otp_dl_demo(path_data)                   # Download the demo data 
log1 <- otp_build_graph(otp = path_otp, 
                        dir = path_data) # Built the demo data into a graph 
log2 <- otp_setup(otp = path_otp, 
                  dir = path_data)       # Start OTP 
``` 

The whole process should only take a few minutes to run, and then the OTP web interface will launch in your browser. You can use the web interface just like any other journey planner, but to get the results into R you need to connect R to the OTP and then make your requests through R.

```{r, eval = FALSE} 
otpcon <- otp_connect()                 # Connect R to OTP 
route <- otp_plan(otpcon,               # Route between two lon/lat coordinates 
                  fromPlace = c(-1.17502, 50.64590),  
                  toPlace = c(-1.15339, 50.72266),
                  mode = c("WALK","TRANSIT")) 
```

> If you have the `tmap` package installed, you can view the route within RStudio. Notice the use of `sf::st_zm(route)` this is because `tmap` does not currently support XYZ coordinates. 

Reviewer: ```suggestion
If you have the `tmap` package installed, you can view the route within RStudio. Notice the use of `sf::st_zm(route)`. This is because `tmap` does not currently support XYZ coordinates. 
```

Fixed grammar by making 2 sentences

```{r, eval = FALSE} 
library(tmap)                           # Load the tmap package 
tmap_mode("view")                       # Set tmap to interactive viewing 
qtm(sf::st_zm(route), lines.lwd = 3, 
    lines.col = "mode")                 # Plot the route on a map 
``` 

> OTP supports loads more features such as batch routing, isochrones, and geocoding, so we recommend working through the full set of [vignettes](https://docs.ropensci.org/opentripplanner/) before trying to set up your own route planner.

Reviewer: fix sentence "OTP supports loads"
